### PR TITLE
Extracted the http client into a separate interface

### PR DESCRIPTION
**Pull request by aakn**
**Changelog**
* Decoupled the http client building from the FluxRuntimeConnector, extracted the http client POST logic to a separate class
* The apache http client is now build via guice
* Adding a few log statements to find the exact error

### CODE REVIEW

File: client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.flipkart.flux.client.runtime;
+
+public interface FluxHttpClient {
+
+    void postOverHttp(Object dataToPost, String pathSuffix);
```
Comment (shyamakirala) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java]: while returning void works for our current use cases, will face issues when we get the response after doing post. This same API is used by both Client and Task runtime to connect to Flux runtime.

File: client/src/main/java/com/flipkart/flux/client/runtime/FluxRuntimeConnectorHttpImpl.java [hunk 2]
```diff
@@ -19,175 +19,83 @@
...
-            final ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
-            objectMapper.writeValue(byteArrayOutputStream, dataToPost);
-            httpPostRequest.setEntity(new ByteArrayEntity(byteArrayOutputStream.toByteArray(), ContentType.APPLICATION_JSON));
-            httpResponse = closeableHttpClient.execute(httpPostRequest);
-            final int statusCode = httpResponse.getStatusLine().getStatusCode();
-            if (statusCode >= Response.Status.OK.getStatusCode() && statusCode < Response.Status.MOVED_PERMANENTLY.getStatusCode() ) {
-                // all is well, TODO write a trace level log
-            } else {
-                // TODO: log status line here
-                throw new RuntimeCommunicationException("Did not receive a valid response from Flux core");
```
Comment (shyamakirala) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxRuntimeConnectorHttpImpl.java]: Closing the connection here should solve the leak issue with having a thread cleanup expired and idle connections periodically.

File: client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java (hunk 1 above)
Comment (aakn) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java]: What we can do is, add another method which takes the type to be returned as parameter (or can be done via generics as well), and use object mapper to convert the response.

File: client/src/main/java/com/flipkart/flux/client/runtime/FluxRuntimeConnectorHttpImpl.java (hunk 2 above)
Comment (aakn) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxRuntimeConnectorHttpImpl.java]: it does get closed in the finally block.

Comment (aakn) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxRuntimeConnectorHttpImpl.java]: yeah, we can add the connection cleanup code, [which will run in background?] to clean up the close wait connections

File: client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java (hunk 1 above)
Comment (aakn) [inline client/src/main/java/com/flipkart/flux/client/runtime/FluxHttpClient.java]: Will the postOverHttp be used by the task runtime? or does the flux runtime connector get used by the runtime?

**Pull request closed without merging**